.gitignore

```diff
@@ -237,6 +237,7 @@ ModelManifest.xml
 
 # Paket dependency manager
 .paket/paket.exe
+.paket/Paket.Restore.targets
 
```

Comment:
can you gitignore and add `nuget nunit` via paket (perhaps in a `group build` to make sure it's not transitive in the Nuget) rather than including the .dll in the repo please ? I'd really like to avoid including binaries and/or otherwise making things heavy if possibel.

Reply:
I can see that @michalkovy opened a pull request with that. If you accept it, then I will merge it here and add it even for the newly added projects.

Reply:
@mchaloupka there are 3x PRs. You are correct that the Paket and NUnit stuff in that overlaps and covers this issue, but his stuff is reliant on your core work so I'm keen for us to work to get this trimmed down to a minimal impl which meets the goal and allows him to build his stuff from your base. @michalkovy it does seem to make sense to carve off your NUnit / Paket bits right away and then get this ready to merge after that ?

Reply:
Seems `Paket.Restore.targets` should actually be added, not `.gitignore`d, see https://fsprojects.github.io/Paket/faq.html#What-files-should-I-commit

Reply:
Probably that information isn't up to date, see [this documentation issue in Paket](https://github.com/fsprojects/Paket/issues/2840#issuecomment-336080872)

Reply:
@mchaloupka The NUnit via Paket changes have been merged, so it might be useful to merge or rebase now

@michalkovy can your Functional DI examples work based on this ? I have yet to look at the latest of this branch or your actual branch, but it would be great if we can establish common abstractions. hopefully with as little as possible in the core itself.

Reply:
I am not sure that I understand the question, but I will try to reply. I fully reused what Milos did, I just missed one method in the container. Milos has there RegisterInstanceAs<'TInterface>method which calls instances.Add(typeof<'TInterface>, instance). That wasn't good enough for me as I need to be able to replace instance in container, too. Hence I added method RegisterOrReplaceInstanceAsType which calls instances.[t] <- instance. However, when I think about more my implementation actually breaks the concept that you would replace DI container how Milos wanted, we will think it through.

I have forward integrated the Functional DI branch. Half of the code is Milos's DI branch.

Reply:
I am not sure that I understand the question, but I will try to reply. I fully reused what Milos did, I just missed one method in the container. Milos has there RegisterInstanceAs<'TInterface>method which calls instances.Add(typeof<'TInterface>, instance). That wasn't good enough for me as I need to be able to replace instance in container, too. Hence I added method RegisterOrReplaceInstanceAsType which calls instances.[t] <- instance. However, when I think about more my implementation actually breaks the concept that you would replace DI container how Milos wanted, we will think it through.

I have forward integrated the Functional DI branch. Half of the code is Milos's DI branch.

Reply:
@michalkovy Thanks for the response. I did assume you were building on Milos' work.

Regarding merging this, my main concern for any impl is that TickSpec's core should touch `IServiceProvider` and `IDisposable` only.

Then the specific DI container (perhaps a batteries included basic impl can live in core to serve the samples but in principle I should be able to wire in a random version of e.g. Autofac if I want) can should be fed in via the top level of the API.

If we get something together that fits that description, then you guys can both build examples of your approaches which would work against an unmodified TickSpec NuGet.

My overall idea/plan is to get either this PR or a new one to do basic plumbing that's able to cater to both needs and then review the approaches on top of that via separate PRs.

Reply:
I think Milos achieved that but unfortunately I need for the functional DI to work the TickSpec core to be able to change instances in container by the TickSpec core. Or, what I could do is to have a custom container in TickSpec core that I know within the container - but I don't feel that it would be better for custom DI framework. So, one option would be to not use IServiceProvider but have a slightly bigger interface that would also contain the method to change an instance in the container. What do you think?

Reply:
I have not looked at your code but its definitely best to keep the interface from the core to the container minimal. IServiceProvider and IDisposable have clear known semantics that we dont need to write an article about. So I'd suggest adding a function or single method interface to model the exact semantic of how stuff is emitted would be best (or could you model it as a post step Event that something external can Subscribe to ?). Growing a big state ball with lots of proprietary methods is the thing I'm seeking to avoid - if we can do it as discrete extension points, there's far better chance we end up providing something useful in the medium term and/or for as yet unforeseen needs.

@mchaloupka what do you think - is there a point in getting minimal `IDisposable` support and extending that or do you see the value being something broader and more complete like Michal's overall approach ?